File: v0.6.0-20260625_2022/.workbuddy/_test/api_tests/12_finance_project_flow_test.php
<?php
// 资金和项目流转测试脚本 - Session B 阶段 3
// 测试目标：验证应收/收款/关闭流程，项目创建/阶段/施工/跟踪流程

require __DIR__ . '/vendor/autoload.php';

use Illuminate\Support\Facades\Http;

class FinanceProjectFlowTest
{
    private string $baseUrl = 'http://172.20.0.139/api';
    private string $token = '';
    private array $testData = [];
    
    public function run(): void
    {
        echo "🚀 Session B 阶段 3：资金和项目流转测试\n";
        echo "==========================================\n\n";
        
        // 1. 登录获取 token
        if (! $this->login()) {
            echo "❌ 登录失败，测试终止\n";
            return;
        }
        
        // 2. 测试资金流转
        $this->testFinanceFlow();
        
        // 3. 测试项目流转
        $this->testProjectFlow();
        
        echo "\n✅ 所有测试完成\n";
    }
    
    private function login(): bool
    {
        echo "🔑 登录获取 token...\n";
        
        $response = Http::post($this->baseUrl . '/auth/login', [
            'username' => 'admin',
            'password' => 'admin123',
        ]);
        
        if ($response->successful() && $response->json('code') === 0) {
            $this->token = $response->json('data.token');
            echo "  ✅ 登录成功\n";
            return true;
        }
        
        echo "  ❌ 登录失败: " . json_encode($response->json()) . "\n";
        return false;
    }
    
    private function testFinanceFlow(): void
    {
        echo "\n💰 测试资金流转...\n";
        echo "------------------------\n";
        
        $http = Http::withToken($this->token);
        
        // 2.1 创建应收款
        echo "  1️⃣ 创建应收款...\n";
        $receivable = [
            'customer_id' => 1, // 假设客户 ID=1 存在
            'project_id' => null,
            'amount' => 100000,
            'received_amount' => 0,
            'status' => 'pending',
            'received_date' => null,
            'note' => '测试应收款',
        ];
        
        $response = $http->post($this->baseUrl . '/finance/receivables', $receivable);
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            $receivableId = $data['data']['id'] ?? null;
            $this->testData['receivable_id'] = $receivableId;
            echo "    ✅ 应收款创建成功 (ID: {$receivableId})\n";
        } else {
            echo "    ❌ 创建失败: " . json_encode($data) . "\n";
            return;
        }
        
        // 2.2 创建收款记录
        echo "  2️⃣ 创建收款记录...\n";
        $payment = [
            'amount' => 50000,
            'payment_date' => now()->toDateString(),
            'payment_method' => 'bank_transfer',
            'note' => '测试收款',
        ];
        
        $response = $http->post(
            $this->baseUrl . "/finance/receivables/{$receivableId}/payments",
            $payment
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 收款记录创建成功\n";
        } else {
            echo "    ❌ 创建失败: " . json_encode($data) . "\n";
        }
        
        // 2.3 再收一笔，凑满 100000
        echo "  3️⃣ 再创建一笔收款记录（凑满）...\n";
        $payment['amount'] = 50000;
        $response = $http->post(
            $this->baseUrl . "/finance/receivables/{$receivableId}/payments",
            $payment
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 第二笔收款记录创建成功\n";
        } else {
            echo "    ❌ 创建失败: " . json_encode($data) . "\n";
        }
        
        // 2.4 关闭应收款
        echo "  4️⃣ 关闭应收款...\n";
        $response = $http->post(
            $this->baseUrl . "/finance/receivables/{$receivableId}/close"
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 应收款关闭成功\n";
        } else {
            echo "    ❌ 关闭失败: " . json_encode($data) . "\n";
        }
        
        // 2.5 验证应收款状态
        echo "  5️⃣ 验证应收款状态...\n";
        $response = $http->get($this->baseUrl . "/finance/receivables/{$receivableId}");
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            $status = $data['data']['status'] ?? '';
            $receivedAmount = $data['data']['received_amount'] ?? 0;
            echo "    ✅ 状态: {$status}, 已收金额: {$receivedAmount}\n";
        } else {
            echo "    ❌ 查询失败: " . json_encode($data) . "\n";
        }
        
        echo "  ✅ 资金流转测试完成\n";
    }
    
    private function testProjectFlow(): void
    {
        echo "\n🏗️ 测试项目流转...\n";
        echo "------------------------\n";
        
        $http = Http::withToken($this->token);
        
        // 3.1 创建项目
        echo "  1️⃣ 创建项目...\n";
        $project = [
            'name' => '测试项目 ' . now()->format('Y-m-d H:i:s'),
            'customer_id' => 1,
            'type' => 'implementation',
            'status' => 'pending',
            'stage' => 'initiation',
            'progress' => 0,
            'member_ids' => [1], // 假设用户 ID=1 存在
        ];
        
        $response = $http->post($this->baseUrl . '/projects', $project);
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            $projectId = $data['data']['id'] ?? null;
            $this->testData['project_id'] = $projectId;
            echo "    ✅ 项目创建成功 (ID: {$projectId})\n";
        } else {
            echo "    ❌ 创建失败: " . json_encode($data) . "\n";
            return;
        }
        
        // 3.2 更新项目阶段（发起 → 询价）
        echo "  2️⃣ 更新项目阶段 (initiation → inquiry)...\n";
        $response = $http->put(
            $this->baseUrl . "/projects/{$projectId}/stage",
            ['stage' => 'inquiry']
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 阶段更新成功\n";
        } else {
            echo "    ❌ 更新失败: " . json_encode($data) . "\n";
        }
        
        // 3.3 继续更新阶段（询价 → 合同）
        echo "  3️⃣ 继续更新阶段 (inquiry → contract)...\n";
        $response = $http->put(
            $this->baseUrl . "/projects/{$projectId}/stage",
            ['stage' => 'contract']
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 阶段更新成功\n";
        } else {
            echo "    ❌ 更新失败: " . json_encode($data) . "\n";
        }
        
        // 3.4 创建施工日志
        echo "  4️⃣ 创建施工日志...\n";
        $log = [
            'content' => '测试施工日志',
            'log_date' => now()->toDateString(),
            'progress' => 30,
        ];
        
        $response = $http->post(
            $this->baseUrl . "/projects/{$projectId}/construction-logs",
            $log
        );
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 施工日志创建成功\n";
        } else {
            echo "    ❌ 创建失败: " . json_encode($data) . "\n";
        }
        
        // 3.5 查看项目跟踪
        echo "  5️⃣ 查看项目跟踪...\n";
        $response = $http->get($this->baseUrl . "/projects/{$projectId}/tracking");
        $data = $response->json();
        
        if ($response->successful() && ($data['code'] ?? -1) === 0) {
            echo "    ✅ 项目跟踪数据获取成功\n";
            echo "    当前阶段: " . ($data['data']['current_stage'] ?? 'unknown') . "\n";
        } else {
            echo "    ❌ 获取失败: " . json_encode($data) . "\n";
        }
        
        echo "  ✅ 项目流转测试完成\n";
    }
}

$test = new FinanceProjectFlowTest();
$test->run();

File: v0.6.0-20260625_2022/.workbuddy/_test/api_tests/12_finance_project_flow_test.py
#!/usr/bin/env python3
"""
Session B 阶段 3：资金和项目流转测试
测试目标：
1. 资金流转：创建应收款 → 创建收款 → 关闭应收款
2. 项目流转：创建项目 → 更新阶段 → 创建施工日志 → 查看跟踪
"""

import requests
import json
from datetime import datetime

BASE_URL = "http://172.20.0.139/api"
TOKEN = None

def login():
    """登录获取 token"""
    print("🔑 登录获取 token...")
    resp = requests.post(f"{BASE_URL}/auth/login", json={
        "username": "admin",
        "password": "admin123",
    }, timeout=10)
    
    if resp.status_code == 200:
        data = resp.json()
        if data.get("code") == 0:
            global TOKEN
            TOKEN = data["data"]["token"]
            print(f"  ✅ 登录成功 (token 前 20 字符: {TOKEN[:20]}...)")
            return True
    
    print(f"  ❌ 登录失败: {resp.status_code} {resp.text[:200]}")
    return False

def auth_request(method, path, **kwargs):
    """带认证的请求"""
    headers = kwargs.pop("headers", {})
    headers["Authorization"] = f"Bearer {TOKEN}"
    headers["Accept"] = "application/json"
    return requests.request(method, f"{BASE_URL}{path}", headers=headers, **kwargs)

def test_finance_flow():
    """测试资金流转"""
    print("\n💰 测试资金流转...")
    print("-" * 40)
    
    # 1. 创建应收款
    print("  1️⃣ 创建应收款...")
    receivable = {
        "customer_id": 1,
        "project_id": None,
        "amount": 100000,
        "received_amount": 0,
        "status": "pending",
        "received_date": None,
        "note": f"测试应收款 {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
    }
    
    resp = auth_request("POST", "/finance/receivables", json=receivable)
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        receivable_id = data["data"]["id"]
        print(f"    ✅ 应收款创建成功 (ID: {receivable_id})")
    else:
        print(f"    ❌ 创建失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
        return
    
    # 2. 创建第一笔收款
    print("  2️⃣ 创建第一笔收款 (50000)...")
    payment = {
        "amount": 50000,
        "payment_date": datetime.now().strftime("%Y-%m-%d"),
        "payment_method": "bank_transfer",
        "note": "测试收款 - 第一笔",
    }
    
    resp = auth_request("POST", f"/finance/receivables/{receivable_id}/payments", json=payment)
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 第一笔收款创建成功")
    else:
        print(f"    ❌ 创建失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 3. 创建第二笔收款（凑满）
    print("  3️⃣ 创建第二笔收款 (50000)...")
    payment["amount"] = 50000
    payment["note"] = "测试收款 - 第二笔"
    
    resp = auth_request("POST", f"/finance/receivables/{receivable_id}/payments", json=payment)
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 第二笔收款创建成功")
    else:
        print(f"    ❌ 创建失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 4. 关闭应收款
    print("  4️⃣ 关闭应收款...")
    resp = auth_request("POST", f"/finance/receivables/{receivable_id}/close")
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 应收款关闭成功")
    else:
        print(f"    ❌ 关闭失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 5. 验证应收款状态
    print("  5️⃣ 验证应收款状态...")
    resp = auth_request("GET", f"/finance/receivables/{receivable_id}")
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        status = data["data"].get("status", "")
        received_amount = data["data"].get("received_amount", 0)
        print(f"    ✅ 状态: {status}, 已收金额: {received_amount}")
        
        if status == "closed" and received_amount == 100000:
            print(f"    ✅ 资金流转测试通过！")
        else:
            print(f"    ⚠️ 状态或金额不正确")
    else:
        print(f"    ❌ 查询失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    print("  ✅ 资金流转测试完成\n")

def test_project_flow():
    """测试项目流转"""
    print("\n🏗️ 测试项目流转...")
    print("-" * 40)
    
    # 1. 创建项目
    print("  1️⃣ 创建项目...")
    project = {
        "name": f"测试项目 {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
        "customer_id": 1,
        "type": "implementation",
        "status": "pending",
        "stage": "initiation",
        "progress": 0,
        "member_ids": [1],
    }
    
    resp = auth_request("POST", "/projects", json=project)
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        project_id = data["data"]["id"]
        print(f"    ✅ 项目创建成功 (ID: {project_id})")
    else:
        print(f"    ❌ 创建失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
        return
    
    # 2. 更新项目阶段（initiation → inquiry）
    print("  2️⃣ 更新项目阶段 (initiation → inquiry)...")
    resp = auth_request("PUT", f"/projects/{project_id}/stage", json={"stage": "inquiry"})
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 阶段更新成功")
    else:
        print(f"    ❌ 更新失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 3. 继续更新阶段（inquiry → contract）
    print("  3️⃣ 继续更新阶段 (inquiry → contract)...")
    resp = auth_request("PUT", f"/projects/{project_id}/stage", json={"stage": "contract"})
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 阶段更新成功")
    else:
        print(f"    ❌ 更新失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 4. 创建施工日志
    print("  4️⃣ 创建施工日志...")
    log = {
        "content": f"测试施工日志 {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}",
        "log_date": datetime.now().strftime("%Y-%m-%d"),
        "progress": 30,
    }
    
    resp = auth_request("POST", f"/projects/{project_id}/construction-logs", json=log)
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        print(f"    ✅ 施工日志创建成功")
    else:
        print(f"    ❌ 创建失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    # 5. 查看项目跟踪
    print("  5️⃣ 查看项目跟踪...")
    resp = auth_request("GET", f"/projects/{project_id}/tracking")
    data = resp.json()
    
    if resp.status_code == 200 and data.get("code") == 0:
        current_stage = data["data"].get("current_stage", "unknown")
        print(f"    ✅ 项目跟踪数据获取成功")
        print(f"    当前阶段: {current_stage}")
    else:
        print(f"    ❌ 获取失败: {resp.status_code} {json.dumps(data, ensure_ascii=False)}")
    
    print("  ✅ 项目流转测试完成\n")

def main():
    print("🚀 Session B 阶段 3：资金和项目流转测试")
    print("=" * 50)
    print()
    
    # 登录
    if not login():
        print("❌ 登录失败，测试终止")
        return
    
    # 测试资金流转
    try:
        test_finance_flow()
    except Exception as e:
        print(f"❌ 资金流转测试异常: {e}")
    
    # 测试项目流转
    try:
        test_project_flow()
    except Exception as e:
        print(f"❌ 项目流转测试异常: {e}")
    
    print("✅ 所有测试完成")

if __name__ == "__main__":
    main()
